account_invoice_supplier_self_invoice/models/account_move.py

```diff
@@ -0,0 +1,87 @@
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    self_invoice_number = fields.Char(readonly=True, copy=False)
+    is_self_invoice_number_different = fields.Boolean(
+        string="Different invoice reference and self invoice number",
+        compute="_compute_is_self_invoice_number_different",
+        help="Self invoice number is different than invoice reference",
+    )
+    set_self_invoice = fields.Boolean(
+        help="Set self invoice for this invoice.",
+    )
+    can_self_invoice = fields.Boolean(related="partner_id.self_invoice")
+
+    def _compute_is_self_invoice_number_different(self):
+        self.update({"is_self_invoice_number_different": False})
+        for move in self:
+            if move.move_type not in ("in_invoice", "in_refund") or not (
+                move.ref and move.self_invoice_number
+            ):
+                continue
+            if move.ref != move.self_invoice_number:
+                move.is_self_invoice_number_different = True
+
+    @api.onchange("partner_id", "company_id")
+    def _onchange_partner_id(self):
+        res = super()._onchange_partner_id()
+        self.set_self_invoice = self.partner_id.self_invoice
+        return res
+
+    def action_post(self):
+        in_move_types = ("in_invoice", "in_refund")
+        # Set today for invoice date in self invoices
+        self.filtered(
+            lambda inv: inv.move_type in in_move_types
+            and inv.set_self_invoice
+            and not inv.invoice_date
+        ).invoice_date = fields.Date.today()
+        res = super().action_post()
+        for invoice in self:
+            partner = invoice.partner_id
+            if (
+                partner.self_invoice
+                and invoice.move_type in in_move_types
+                and invoice.set_self_invoice
+                and not invoice.self_invoice_number
+            ):
+                sequence = partner.self_invoice_sequence_id
+                invoice.self_invoice_number = sequence.with_context(
+                    ir_sequence_date=invoice.date
+                ).next_by_id()
+                invoice.ref = invoice.self_invoice_number
+        return res
+
+    @api.model
+    def create(self, vals):
+        onchanges = {
+            "_onchange_partner_id": ["set_self_invoice"],
+        }
+        if vals.get("move_type") in ("in_invoice", "in_refund"):
+            for onchange_method, changed_fields in onchanges.items():
+                if any(f not in vals for f in changed_fields):
+                    invoice = self.new(vals)
+                    getattr(invoice, onchange_method)()
+                    for field in changed_fields:
+                        if field not in vals and invoice[field]:
+                            vals[field] = invoice._fields[field].convert_to_write(
+                                invoice[field], invoice
+                            )
+        return super().create(vals)
```

Comment:
This code is... weird. Also I'd say it won't perform very well. You can replace it with a simple compute:

```suggestion
    @api.depends("partner_id", "company_id")
    def _compute_self_invoice(self):
        """Inherit default for self-invoice from partner."""
        for move in self:
            self.set_self_invoice = self.partner_id.self_invoice
```

Make sure you make the field computed, stored and writable. Also remove the `_onchange_partner_id()` override above.